pkg/plugin/plugin.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"path"
+	"strings"
+
+	"github.com/spf13/pflag"
+
+	"sigs.k8s.io/kubebuilder/pkg/model/config"
+)
+
+const DefaultNameQualifier = ".kubebuilder.io"
+
+type Base interface {
+	// Name returns a DNS1123 label string defining the plugin type.
+	// For example, Kubebuilder's main plugin would return "go".
+	Name() string
+	// Version returns the plugin's semantic version, ex. "v1.2.3".
+	//
+	// Note: this version is different from config version.
+	Version() string
+	// SupportedProjectVersions lists all project configuration versions this
+	// plugin supports, ex. []string{"2", "3"}. The returned slice cannot be empty.
+	SupportedProjectVersions() []string
+}
+
+// Key returns a unique identifying string for a plugin's name and version.
+func Key(name, version string) string {
```

Comment:
nit: just do `type Key string` and make some of the split, etc be methods on that?

Reply:
Followup?

Reply:
sure